sycl/doc/extensions/experimental/sycl_ext_oneapi_matrix/sycl_ext_intel_matrix.asciidoc

```diff
@@ -21,135 +19,403 @@
+partitioning among WIs is implementation defined.
+
+2- Sub group size is not generally fixed.
+
+The code listing below shows a synopsis of these new APIs.
+
+```c++
+namespace sycl::ext::intel::experimental::matrix {
+
+wi_data<group, T, Use, Rows, Cols, Layout> get_wi_data(Group g,
+ joint_matrix<Group, T, Use, Rows, Cols, Layout> Mat);
+
+template <typename T, size_t Rows, size_t Cols, use Use, layout
+Layout, typename Group>
+class wi_data {
+  size_t length();
+  wi_element<T, Rows, Cols, Use, Layout, Group> operator[](size_t i);
+};
+template <typename T, size_t Rows, size_t Cols,
+          use Use, layout Layout,
+          typename Group = sycl::sub_group>
+class wi_element {
+  operator T();
+  wi_element &operator=(const T &rhs);
+  wi_element &operator+=(const T &rhs);
+  wi_element &operator-=(const T &rhs);
+  wi_element &operator*=(const T &rhs);
+  wi_element &operator/=(const T &rhs);
+
+  std::tuple<size_t, size_t> get_coord();
+};
+
+} // namespace sycl::ext::intel::experimental::matrix
+```
+
+In the following example `wi_data_c` is a reference to the WI owned
+portion of the joint matrix `matC`. As such `wi_data_c[i] OP rhs`
+updates the corresponding matrix element in the joint_matrix `matC`.
+Vectorization along the sub group dimension will get enabled
+automatically to vectorize the contiguous portion of the matrix.
+
+
+```c++
+auto wi_data_c = get_wi_data(sg, matC);
+for (int i = 0; i < wi_data_c.length(); i++)
+  wi_data_c[i] *= alpha;    // Note that the indexing here "i"
+  //is in the vector owned by a WI, not in the matrix C
+```
+
+===== Work-item data to joint matrix mapping coordinates
+The `wi_data` and `wi_element` classes provide access to the matrix
+elements that are local to the calling work-item. However, the
+distribution of matrix elements to each work-item is
+implementation-defined, so application code cannot assume any fixed
+distribution. Instead, application code can use the `get_coord` method
+to query the matrix coordinates of an individual `wi_element`.
+
+`get_coord` returns [row,col] coordinates of the current object
+`wi_element` of the joint matrix.  The code above results into the following:
+
+```c++
+auto data = get_wi_data(sg, tA);
+// each WI calculates local sum of rows
+for (int i = 0; i < data.length(); ++i) {
+  auto [row, col] = data[i].get_coord();
+  sum_local_rows[row] += data[i];
 }
 ```
 
+===== Extending `joint_matrix_apply` with the mapping coordinates
+For use cases in which the order of accessing the elements of
+`wi_data` is important, indexing and the mapping coordinates API
+provided above should be used. If access order is not important,
+`joint_matrix_apply` is extended to provide the `row` and `col`
+arguments as part of the function object.
+
+Consequently, the sum of rows example provided above can be written as
+follows:
+
+```c++
+joint_matrix_apply(sg, A, [=](T &val, size_t row, size_t  col) {
+   sum_local_rows[row] += val;
+});
+```
```

Comment:
You should make it clear that this new form of `joint_matrix_apply` is in the `ext::intel::matrix` namespace.  For example:

> === Per-element operations with coordinates
>
> This extension adds a new form of the `joint_matrix_apply` function in the `sycl::ext::intel::matrix` namespace that allows the application to perform an operation on each element of the matrix.  This function is similar to the form in sycl_ext_oneapi_joint_matrix, but it also provides the matrix coordinates of each element to the callback function:
>
> ```
> namespace sycl::ext::intel::experimental::matrix {
>
> template<typename Group, typename T, use Use, size_t Rows, size_t Cols, layout Layout, typename F>
> void joint_matrix_apply(Group g, joint_matrix<Group, T, Use, Rows, Cols, Layout>& C, F&& func);
>
> } // namespace sycl::ext::intel::experimental::matrix
> ```
>
> The `func` callback is invoked with three parameters `(T& element, size_t row, size_t col)`, where `row` and `col` tell the coordinates of `element` in the joint matrix.  To illustrate, the following example shows how you can use this API to sum the rows of a matrix:
>
> [example]